Remove dead code in SavedVariable

Varal7 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59838 Remove dead code in SavedVariable**
* #59837 Avoid un-necessary unwrapping of Tensor in SavedVariable

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29069214](https://our.internmc.facebook.com/intern/diff/D29069214)

Varal7 committed:
Remove dead code in SavedVariable

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Remove dead code in SavedVariable"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Remove dead code in SavedVariable"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/saved_variable.h:
```diff
@@ -42,7 +42,7 @@ class TORCH_API SavedVariable {
 
   void reset_grad_function() {
-    grad_fn_.reset();
+    // No op
```
Is this function ever called?

Varal7 replied:
Its used here: https://github.com/pytorch/pytorch/blob/master/tools/autograd/gen_autograd_functions.py#L327

I guess I can remove it in this PR as well?

albanD reviewed (commented):
code removal! 

It would be nice to do a small benchmark of the impact of this stack on perf.
In particular, using the Timer class with instruction count. We can chat about this offline if you want.

Varal7 commented:
> code removal!
> 
> It would be nice to do a small benchmark of the impact of this stack on perf.
> In particular, using the Timer class with instruction count. We can chat about this offline if you want.

Thanks! As a spoiler, I already ran the following test:
```
# Benchmark from: https://github.com/pytorch/pytorch/pull/56017#issuecomment-853388987

import torch
import time

x = torch.zeros(1, device='cpu', requires_grad=True)
for _ in range(100000):
    x = x * x

start = time.time()
x.backward()
end = time.time()
print(end-start)

# Without the patch:

➜ for i in `seq 1 10`; do python bench.py; done
0.9253489971160889
0.9564845561981201
0.9363410472869873
0.897282600402832
0.8867707252502441
0.8611898422241211
0.8699934482574463
0.8362593650817871
0.8579938411712646
0.8739769458770752

# With the patch:

➜ for i in `seq 1 10`; do python bench.py; done
0.5738115310668945
0.5449838638305664
0.5333981513977051
0.5343718528747559
0.5725717544555664
0.5140812397003174
0.5454022884368896
0.5519816875457764
0.531273365020752
0.5399165153503418

```

Varal7 commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -96,8 +88,8 @@ Variable SavedVariable::unpack(std::shared_ptr<Node> saved_for) const {
-                                                   : grad_fn_;
-  if (has_grad_fn_ && !grad_fn) {
+                                                   : nullptr;
+  if (!saved_original && !grad_fn) {
```
The logic is:
If you saved the original variable, you never want to look at what's in `saved_for`.
Now if you did not save the original variable, it can not be a leaf, so `has_grad_fn` is always true.

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29069214).

Varal7 committed:
Update on "Remove dead code in SavedVariable"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29069214](https://our.internmc.facebook.com/intern/diff/D29069214)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29069214).

Varal7 commented:
Benchmark!

## Wall time

```python
# bench.py
x = torch.zeros(1, device='cpu', requires_grad=True)

start_fw = time.time()
for _ in range(100000):
    x = x * x
end_fw = time.time()

start_bw = time.time()
x.backward()
end_bw = time.time()
print("Forward\t", end_fw-start_fw, end="\t")
print("Backward\t", end_bw-start_bw)
```

Without the patch:

```
➜ for i in `seq 1 10`; do python bench.py; done
Forward  0.47095823287963867    Backward         0.9542675018310547
Forward  0.388355016708374      Backward         0.9605531692504883
Forward  0.42574310302734375    Backward         1.0612819194793701
Forward  0.38661670684814453    Backward         0.9773132801055908
Forward  0.37992262840270996    Backward         0.9450616836547852
Forward  0.3741035461425781     Backward         0.8979041576385498
Forward  0.423450231552124      Backward         1.049081802368164
Forward  0.39497900009155273    Backward         0.9403750896453857
Forward  0.3828461170196533     Backward         0.9505488872528076
Forward  0.36185240745544434    Backward         0.9551246166229248
```

With the patch:

```

➜ for i in `seq 1 10`; do python bench.py; done
Forward  0.33322834968566895    Backward         0.6979451179504395
Forward  0.34907960891723633    Backward         0.7061493396759033
Forward  0.34142494201660156    Backward         0.6910700798034668
Forward  0.34435534477233887    Backward         0.6630330085754395
Forward  0.3300626277923584     Backward         0.701082706451416
Forward  0.4510471820831299     Backward         0.874150276184082
Forward  0.3268928527832031     Backward         0.7102131843566895
Forward  0.3682827949523926     Backward         0.6823954582214355
Forward  0.3338043689727783     Backward         0.6908135414123535
Forward  0.3473036289215088     Backward         0.7074475288391113
```

## Instruction count

100 runs per measurement, 1 thread

### Forward

```
stmt: y = (x * x).sum()
setup: x = torch.randn(5, requires_grad=True)
```

Without the patch:

```
                           All          Noisy symbols removed
    Instructions:      2507869                    2492089
    Baseline:              899                        801
```

With the patch:

```
                           All          Noisy symbols removed
    Instructions:      2242869                    2214453
    Baseline:              899                        801
```


### Backward

```
stmt: y.backward(retain_graph=True)
setup: x = torch.randn(5, requires_grad=True); y = (x * x).sum()
```

Without the patch:
```
                           All          Noisy symbols removed
    Instructions:      7286245                    7178227
    Baseline:              919                        801
```

With the patch:

```
                           All          Noisy symbols removed
    Instructions:      6939964                    6847261
    Baseline:              899                        801
```

albanD commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -131,30 +123,20 @@ Variable SavedVariable::unpack(std::shared_ptr<Node> saved_for) const {
     return data_;
   }
 
+  // From now on, we can assume the variable is not a leaf and is an output.
```
nit: maybe also say that we have a valid grad_fn that should be attached to the returned Tensor.

albanD reviewed (approved):
Code simplification is always nice!! :)

Varal7 committed:
Update on "Remove dead code in SavedVariable"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29069214](https://our.internmc.facebook.com/intern/diff/D29069214)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29069214).

Varal7 committed:
Update on "Remove dead code in SavedVariable"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29069214](https://our.internmc.facebook.com/intern/diff/D29069214)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29069214).
